improve scheduler

Spacechild1 opened the pull request:
General:
* `glob_exit` breaks from the scheduler loop, instead of calling `exit`, and thus allows for proper cleanup in `sys_main`.
* `sys_main` calls a new function `pd_term` (counterpart to `pd_init`) before returning. Currently, `pd_term` just destroys the root canvases, but LATER it should free *all* resources. Ideally, Pd should not leak any memory on shutdown.
* fix `audio_shouldkeepopen` for the case where no audio backend has actually been opened (yet). Now `pd -jack` immediately connects to the Jack server; before you had to turn on DSP first.
* if the user switches backends or applies audio settings, only open audio if DSP is on or audio should be kept open
* only derive sleep grain from `sys_schedadvance` if we are in the polling scheduler and DSP is turned on, otherwise use a fixed (default) sleep grain of 1 ms.

Callback scheduler:
* allow to toggle DSP
* allow to switch to the polling scheduler
* check if audio device got stuck
* fix high CPU usage with portaudio

Polling scheduler:
* don't sleep if we already slept in `sys_send_dacs` (= `SENDDACS_SLEPT`)

Jack/portaudio
* enable `THREADSIGNAL` by default
* use a semaphore instead of a condition variable to get rid of unnecessary locking
* only sleep on the semaphore when really idle

Portaudio
* rewrote `pa_send_dacs` to match the much simpler `jack_send_dacs`
* show audio I/O error if we have been late and the audio callback had to drop samples (just like the Jack backend already does)
* remove unused code (`FAKEBLOCKING` not defined) to improve readability

Jack
* gracefully handle server shutdown
* log Jack errors with `PD_DEBUG` to avoid spamming the console

MIDI
* only delay outgoing MIDI messages if audio is delayed (= polling scheduler with DSP on)

---

`sys_reopen_audio` and `sys_close_audio` now only notify the scheduler, therefore they can be safely called anywhere, including the audio callback. The actual implementation has been moved to *private* functions (`sys_do_reopen_audio` and `sys_do_close_audio`)

At least here on Windows, the (new) `THREADSIGNAL` implementation allows for significantly lower latency.

- Closes https://github.com/pure-data/pure-data/issues/550
- Closes https://github.com/pure-data/pure-data/issues/1554 
- Closes https://github.com/pure-data/pure-data/issues/1703
- Closes https://github.com/pure-data/pure-data/issues/1930

This branch has been successfully tested with the following configurations:

* Windows / portaudio (ASIO) / polling scheduler + callback scheduler
* macOS / portaudio (CoreAudio) / polling scheduler + callback scheduler
* macOS / Jack / polling scheduler + callback scheduler
* Linux / Jack / polling scheduler + callback scheduler

Spacechild1 commented:
I have been using this branch for several weeks in an opera project with no crashes and no dropouts. My setup was: Thinkpad P14s, Windows 10, RME MADI-Face, 32 input channels, 36 output channels, 256 samples block size, 10 ms latency.

I will make some quick tests on macOS (portaudio) and Linux (Jack), but since I'm mostly running Pd on Windows, it would be great if others (@umlaeute, @porres, @danomatika, @reduzent) could also give it a try.

danomatika commented:
Here's a Pd 0.53-1 arm64 build for newer mac systems which includes the updated portaudio 19.7.0 and the contents of this branch. It opens and produces audio for the test patch on my macOS 11.7 system. I have not texted further yet.

http://docs.danomatika.com/pdbuilds/0.53/Pd-0.53-1-arm64-schedulerfix.zip

Spacechild1 commented:
@danomatika Thanks!

> It opens and produces audio for the test patch on my macOS 11.7 system

That's good news :-)

danomatika commented:
@millerpuckette This may be a *tentative* fix for the macOS 13 issue. It probably needs to be looked over and tested on Linux.

danomatika commented:
Also @millerpuckette , to test this I pulled Christof's branch via:

    git fetch origin pull/1756/head:scheduler_fix
    git checkout scheduler_fix
    git merge master

This merged cleanly, then I could build with the `scheduler_fix` branch changes applied.

millerpuckette commented:
This is promising but... am I not right that just updating portaudio was enough to fix the Pd hanging problem on MACOS 13?  If so I should imiedialtely push that one out and hold off on this one for a "feature" release.

Spacechild1 commented:
> am I not right that just updating portaudio was enough to fix the Pd hanging problem on MACOS 13?

I think so. This PR is not specific to macOS 13. It just happened that one user was accidentally using the "callbacks" option and wondered why Pd got stuck when stopping DSP. I just mentioned as a side note that this PR would fix all the "callback" related problems - which are not confined to macOS.

> and hold off on this one for a "feature" release.

Definitely!

Spacechild1 commented:
@millerpuckette As a side note: may I go ahead and remove the non-`FAKEBLOCKING` parts in `s_audio_pa.c` (`Pa_WriteStream` and `Pa_ReadStream`)? They have never been used (and most likely never will) and just clutter the code...

Spacechild1 commented:
Seems like it works fine on all configurations. I have removed the [WIP] label and updated the PR description.

reduzent commented:
> It just happened that one user was accidentally using the "callbacks" option and wondered why Pd got stuck when stopping DSP.

I was using callbacks on purpose and wondered why Pd uses 100% of a core (not knowing that those two are related). This seems to be specific for CoreAudio/maOS. There is no callback option in ALSA and on JACK there is no problem with using callbacks.

Spacechild1 commented:
> This seems to be specific for CoreAudio/maOS. 

It was really a problem with Pd's portaudio backend (`s_audio_pa.c`). I also got the high CPU usage when using ASIO on Windows with "callbacks" enabled.

(For others: this issue is also fixed with this PR.)

reduzent commented:
I found some crasher conditions that I didn't spot before, either because of sloppy testing or because things have changed.  This is on Ubuntu 22.04 with current scheduler_fix (b5b448f0e5).

### Switching from Jack with callbacks to Jack without callbacks

Switching to using callbacks seems to always work, but when I uncheck 'Use callbacks' in the audio settings dialog and hit [Apply]  a few times, I can reliably crash Pd. Here is the stack trace:

```
gdb -ex=run --args pd -jack -channels 2 -callback
```

```
Thread 16 "pd" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffff6c7e640 (LWP 372388)]
__new_sem_post (sem=0x0) at ./nptl/sem_post.c:35
35	./nptl/sem_post.c: No such file or directory.
(gdb) btwatchdog: signaling pd...

#0  __new_sem_post (sem=0x0) at ./nptl/sem_post.c:35
#1  0x0000555555682f89 in sys_semaphore_post (sem=<optimized out>) at s_audio_paring.c:382
#2  0x000055555567c2ca in jack_polling_callback (nframes=128, unused=<optimized out>) at s_audio_jack.c:114
#3  0x00007ffff7d6ab0f in ?? () from /lib/x86_64-linux-gnu/libjack.so.0
#4  0x00007ffff7d6a7d8 in ?? () from /lib/x86_64-linux-gnu/libjack.so.0
#5  0x00007ffff7d87910 in ?? () from /lib/x86_64-linux-gnu/libjack.so.0
#6  0x00007ffff7a94b43 in start_thread (arg=<optimized out>) at ./nptl/pthread_create.c:442
#7  0x00007ffff7b26a00 in clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:81
 ```

BTW: I'm using JACK from Ubuntu which probably isn't compiled with debug symbols. Do you need this?

### Switching to ALSA always crashes

When running Pd with JACK, I cannot switch to ALSA. Steps to reproduce the crash:

1. run: `gdb -ex=run --args pd -jack -channels 2 -callback`
2. Go to `Media` -> `ALSA`
3. Hit `Apply`

Backtrace:

```
Program received signal SIGHUP, Hangup.
__futex_abstimed_wait_common64 (private=975940096, cancel=true, abstime=0x7fffffffdce0, op=393, expected=0, futex_word=0x5555556bfe28 <sched_cond+40>) at ./nptl/futex-internal.c:57
57	./nptl/futex-internal.c: No such file or directory.
(gdb) 
(gdb) watchdog: signaling pd...
btwatchdog: signaling pd...

#0  __futex_abstimed_wait_common64 (private=975940096, cancel=true, abstime=0x7fffffffdce0, op=393, expected=0, 
    futex_word=0x5555556bfe28 <sched_cond+40>) at ./nptl/futex-internal.c:57
#1  __futex_abstimed_wait_common (cancel=true, private=975940096, abstime=0x7fffffffdce0, clockid=7442, expected=0, 
    futex_word=0x5555556bfe28 <sched_cond+40>) at ./nptl/futex-internal.c:87
#2  __GI___futex_abstimed_wait_cancelable64 (futex_word=futex_word@entry=0x5555556bfe28 <sched_cond+40>, expected=expected@entry=0, 
    clockid=clockid@entry=0, abstime=abstime@entry=0x7fffffffdce0, private=private@entry=0) at ./nptl/futex-internal.c:139
#3  0x00007ffff7a93f1b in __pthread_cond_wait_common (abstime=0x7fffffffdce0, clockid=0, mutex=0x5555556bfdc0 <sched_mutex>, 
    cond=0x5555556bfe00 <sched_cond>) at ./nptl/pthread_cond_wait.c:503
#4  ___pthread_cond_timedwait64 (cond=cond@entry=0x5555556bfe00 <sched_cond>, mutex=mutex@entry=0x5555556bfdc0 <sched_mutex>, 
    abstime=abstime@entry=0x7fffffffdce0) at ./nptl/pthread_cond_wait.c:652
#5  0x00005555555feb34 in m_callbackscheduler () at m_sched.c:474
#6  m_mainloop () at m_sched.c:516
#7  0x0000555555608f45 in sys_main (argc=5, argv=<optimized out>) at s_main.c:393
#8  0x00007ffff7a29d90 in __libc_start_call_main (main=main@entry=0x55555556e6a0 <main>, argc=argc@entry=5, argv=argv@entry=0x7fffffffdeb8)
    at ../sysdeps/nptl/libc_start_call_main.h:58
#9  0x00007ffff7a29e40 in __libc_start_main_impl (main=0x55555556e6a0 <main>, argc=5, argv=0x7fffffffdeb8, init=<optimized out>, 
    fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffdea8) at ../csu/libc-start.c:392
#10 0x000055555556e6d5 in _start ()

```

Spacechild1 commented:
> Switching to ALSA always crashes
> When running Pd with JACK, I cannot switch to ALSA. Steps to reproduce the crash:

AFAICT, the actual issue is trying to switch from an API with callbacks *enabled* to another API that does not support callbacks. I had the same issue trying to switch from ASIO (with callbacks) to MMIO. This is caused by a smal oversight in `glob_audio_setapi`. I just pushed a fix, can you give it a try?

Spacechild1 commented:
> Switching from Jack with callbacks to Jack without callbacks

> Switching to using callbacks seems to always work, but when I uncheck 'Use callbacks' in the audio settings dialog and hit [Apply] **a few times**, I can reliably crash Pd. 

Just to be sure: does "a few times" only refer to "hit [Apply]", or to the whole sequence? Also, are you sure that the crash really happens when going from "callbacks" to "no callbacks" - and not the other way around?

reduzent commented:
> > Switching to using callbacks seems to always work, but when I uncheck 'Use callbacks' in the audio settings dialog and hit [Apply] **a few times**, I can reliably crash Pd.
> 
> Just to be sure: does "a few times" only refer to "hit [Apply]", or to the whole sequence of actions?

Once the checkbox is unchecked, I can crash it by hitting `Apply` once or multiple times.

reduzent commented:
Here a report from macOS 12.6.3 (amd64) with b5b448f0e5007. Steps to reproduce:

1. Start Pd in debugger: ` lldb -- /Users/user/Desktop/Pd-scheduler_fix.app/Contents/Resources/bin/pd -jack -channels 2 -callback`
2. Open Dialog `Media` -> `jack`
3. Uncheck `Use callbacks`
4.  Hit `Apply`

I immediately get: 

```
Process 8855 stopped
* thread #7, stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
    frame #0: 0x00007ff81d293882 libsystem_pthread.dylib`pthread_mutex_lock + 4
libsystem_pthread.dylib`pthread_mutex_lock:
->  0x7ff81d293882 <+4>:  cmpq   $0x4d55545a, (%rdi)       ; imm = 0x4D55545A
    0x7ff81d293889 <+11>: jne    0x7ff81d2938fe            ; <+128>
    0x7ff81d29388b <+13>: movl   0xc(%rdi), %eax
    0x7ff81d29388e <+16>: movl   %eax, %ecx
Target 0: (pd) stopped.
```

Backtrace:

```
(lldb) bt
* thread #7, stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
  * frame #0: 0x00007ff81d293882 libsystem_pthread.dylib`pthread_mutex_lock + 4
    frame #1: 0x00000001000dd7ae pd`sys_semaphore_post(sem=0x0000000000000000) at s_audio_paring.c:384:5 [opt]
    frame #2: 0x00000001000db249 pd`jack_polling_callback(nframes=64, unused=<unavailable>) at s_audio_jack.c:114:5 [opt]
    frame #3: 0x00000001005ed975 libjack.0.1.0.dylib`___lldb_unnamed_symbol437 + 229
    frame #4: 0x00000001005ed9aa libjack.0.1.0.dylib`___lldb_unnamed_symbol438 + 10
    frame #5: 0x00000001005fd232 libjack.0.1.0.dylib`___lldb_unnamed_symbol899 + 130
    frame #6: 0x00007ff81d2984e1 libsystem_pthread.dylib`_pthread_start + 125
    frame #7: 0x00007ff81d293f6b libsystem_pthread.dylib`thread_start + 15
```

Spacechild1 commented:
> Once the checkbox is unchecked, I can crash it by hitting Apply once or multiple times.

Aha. Does it really happen when you hit Apply only *once*? Or rather when you hit it multiple times? In case of the latter, I would already have an explanation.

reduzent commented:
> > Once the checkbox is unchecked, I can crash it by hitting Apply once or multiple times.
> 
> Aha. Does it really happen when you hit Apply _once_? Or rather when you hit it multiple times? In case of the latter, I would already have an explanation.

Sometimes it is sufficient to hit `Apply` exactly once (in a fresh run of Pd). More often, it requires more times.

Spacechild1 commented:
> I'm pretty confident (but also not absolutely sure) that this worked before. From what I remember, checking if disabling/enabling callbacks works smoothly was the point of my last testing session.

I does work fine with portaudio here on Windows, so it seems to be a Jack problem (on Linux and macOS) - which is not surprising considering that I have never actually tested the Jack implementation myself :-) Will investigate.

---

Does switching from Jack (with callbacks) to ALSA work now?

Spacechild1 commented:
> Here a report from macOS 12.6.3 (amd64) with https://github.com/pure-data/pure-data/commit/b5b448f0e5007b042d798006f47e285830164b49. Steps to reproduce:

Hmmm, I can't reproduce this on macOS 13.1. I start Pd with `./pd -jack -channels 2 -callbacks` and I can toggle between "callbacks" and "no callbacks" just fine...

Spacechild1 commented:
Actually, b5b448f0e5007b042d798006f47e285830164b49 is outdated. (I have force pushed in the meantime.) Can you try with the current HEAD (f89ae374923a05076d6f5067579b97a76fe986ea)?

reduzent commented:
> Actually, [b5b448f](https://github.com/pure-data/pure-data/commit/b5b448f0e5007b042d798006f47e285830164b49) is outdated. (I have force pushed in the meantime.) Can you try with the current HEAD ([f89ae37](https://github.com/pure-data/pure-data/commit/f89ae374923a05076d6f5067579b97a76fe986ea))?

I get the same crashes with f89ae37. To make sure that I am using JACK in a well-known state, I re-installed this [package](https://github.com/jackaudio/jack2-releases/releases/download/v1.9.22/jack2-macOS-universal-v1.9.22.tar.gz). The current version is 1.9.22. I was probably using 1.9.21 before. Now, I get a similar behavior to Linux: I have to hit `Apply` multiple times sometimes in order to crash Pd.

reduzent commented:
BTW: are you testing on an Intel or Arm Mac?

Spacechild1 commented:
I am using an Intel Mac with the same Jack package you have linked (Jack 1.9.22 Universal installer). Can you verify that the crash does not happen with portaudio (CoreAudio)?

reduzent commented:
Another way to crash Pd (f89ae374923a0) on macOS:

1. Start Pd: `lldb -- /Users/user/Desktop/Pd-scheduler_fix.app/Contents/Resources/bin/pd -pa`
2. Do _not_ turn DSP on
3. Open the Dialog `Media` -> `jack`
4. Crash

```
Process 1907 stopped
* thread #7, stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
    frame #0: 0x00007ff80514f882 libsystem_pthread.dylib`pthread_mutex_lock + 4
libsystem_pthread.dylib`pthread_mutex_lock:
->  0x7ff80514f882 <+4>:  cmpq   $0x4d55545a, (%rdi)       ; imm = 0x4D55545A
    0x7ff80514f889 <+11>: jne    0x7ff80514f8fe            ; <+128>
    0x7ff80514f88b <+13>: movl   0xc(%rdi), %eax
    0x7ff80514f88e <+16>: movl   %eax, %ecx
Target 0: (pd) stopped.
```
Backtrace:

```
(lldb) bt
* thread #7, stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
  * frame #0: 0x00007ff80514f882 libsystem_pthread.dylib`pthread_mutex_lock + 4
    frame #1: 0x00000001000dd67e pd`sys_semaphore_post(sem=0x0000000000000000) at s_audio_paring.c:384:5 [opt]
    frame #2: 0x00000001000db119 pd`jack_polling_callback(nframes=128, unused=<unavailable>) at s_audio_jack.c:114:5 [opt]
    frame #3: 0x00000001005ed975 libjack.0.1.0.dylib`___lldb_unnamed_symbol437 + 229
    frame #4: 0x00000001005ed9aa libjack.0.1.0.dylib`___lldb_unnamed_symbol438 + 10
    frame #5: 0x00000001005fd232 libjack.0.1.0.dylib`___lldb_unnamed_symbol899 + 130
    frame #6: 0x00007ff8051544e1 libsystem_pthread.dylib`_pthread_start + 125
    frame #7: 0x00007ff80514ff6b libsystem_pthread.dylib`thread_start + 15
```

reduzent commented:
Things are different between using Pd's menus and using the mediasettings library. With your branch, Pd doesn't hang anymore when switching backends, even when `use callbacks` is enabled. However, when using the mediasettings' libraries [audiosettings] object to switch backends, Pd still hangs when `callbacks` are enabled. Is there some magic in the Tk menus  to make sure callbacks are off before switching?  If so, should I make sure programming-wise that callbacks get disabled _before_ swtiching backends? I tried that before with Pd from master and I couldn't avoid Pd hanging because it looked like it takes some time before it was safe to switch and I wasn't able to reliably detect how much time was necessary. It still would hang from time to time after switching, even when I waited a few seconds.

Spacechild1 commented:
> However, when using the mediasettings' libraries [audiosettings] object to switch backends, Pd still hangs when callbacks are enabled. 

Can you provide an example patch?

Spacechild1 commented:
@reduzent I finally found the error! I accidentally created the semaphore *after* opening the stream which created a race condition (that was only triggered with very low block sizes).

Now everything should work fine. Can you pull and double check?

reduzent commented:
> @reduzent I finally found the error! 

Sweet!

> I accidentally created the semaphore _after_ opening the stream which created a race condition (that was only triggered with very low block sizes).

I indeed was running jackd with blocksizes of 64 and 128 (setup from latency testing). This shows how many details are important while reporting. 

> Now everything should work fine. Can you pull and double check?

I confirm I can't crash Pd anymore by hammering the `Apply` button in the `Media` -> `jack` dialog on Linux. Will do some more thorough testing this afternoon, also on macOS.

Spacechild1 commented:
Nice! Please also test with the mediasettings library.

reduzent commented:
Ok, I did some more testing and
  * switching between JACK and CoreAudio works fine on macOS
  * switching between JACK and ALSA works fine on Linux
regardless of the current callback setting.

Switching between backends does not respect the previous callback setting, they are always off after switching. It's just an observation, I don't think it is necessarily bad.

reduzent commented:
Hammering `Apply` in the `jack` dialog does not crash anymore.

reduzent commented:
Switching between backends with the [audiosettings] (from [mediasettings](http://deken.puredata.info/library/mediasettings)) external locks Pd, when callbacks are enabled. It seems any audio processing is stopped and Pd doesn't react to GUI interactions. Switching works fine while callbacks are off. I tried to avoid locks by switch off callbacks with each `driver` change with a message like `params @callback 0, driver $1`  but it still locks Pd. 

See attached patch.
[audiosettings-test.zip](https://github.com/pure-data/pure-data/files/10943703/audiosettings-test.zip)

reduzent commented:
@Spacechild1 One other thing: When I load netpd-sessions in the `scheduler_fix` build, I experience some glitches. For instance, One synth goes into fully overblown mode after the third loop, like if an oscillator would be amplified by a large value. I had similar issues when work on multi-channel stuff was going on. Current master is fine, though. Could it be that `scheduler_fix` is not on par with current master regarding mutlichannel changes?  I was reluctant to merge current `master` into your `scheduler_fix` branch because I wanted to make sure we test on the exact same thing and wanted to avoid another dimension of complexity. But maybe it would be good to rebase `scheduler_fix` onto current `master`? Or what do you think?

Spacechild1 commented:
> Switching between backends does not respect the previous callback setting, they are always off after switching. It's just an observation, I don't think it is necessarily bad.

This is expected. When you change the audio backends, most audio settings are reset to the default.

> But maybe it would be good to rebase scheduler_fix onto current master?

Done.

> Switching between backends with the [audiosettings] (from [mediasettings](http://deken.puredata.info/library/mediasettings)) external locks Pd

Thanks, will have a look!

Spacechild1 commented:
The problem is really in the `[audiosettings]` object:

https://git.iem.at/pd/mediasettings/-/blob/master/audiosettings.c#L665-667

This will deadlock when called from within the audio callback, i.e. if callbacks are enabled.

The general problem is that the `[audiosettings]` library makes assumptions about Pd internals. For the `[params(` method it actually just sends a Pd message, which in turn will do the right thing. Unfortunately, it can't do the same for the `[driver(` method because the `"audio-setapi"` message would open the audio dialog. As a workaround, it mimicks the (old) implementation of `glob_audio_setapi`, but this does not work with the callback scheduler. At the moment, it is not possible for `[audiosettings]` to change backends once your in the callback scheduler.

Here's a solution: we could add a second (optional) argument to the `"audio-setapi"` message to surpress the audio dialog. Then the `[audiosettings]` object could use that message and wouldn't have to worry about Pd scheduler internals.

I am curious what @umlaeute thinks.

Spacechild1 commented:
> I tried to avoid locks by switch off callbacks with each driver change with a message like params @callback 0, driver $1 but it still locks Pd.

This does not work because the `[driver ...(` message is still sent from within the audio callback. You need to add some delay so that Pd has a chance to exit the callback scheduler and enter the polling scheduler. Adding a 1 second delay between `[params @callback 0(` and `[driver 3(` seems to do the trick - at least here on Windows when switching from Portaudio with callbacks to MMIO. Needless to say, this is just a workaround and not a real solution.

reduzent commented:
> > But maybe it would be good to rebase scheduler_fix onto current master?
> 
> Done.

Cool. The problem with some session producing "wrong" audio is gone. I'm glad it  was not related to scheduler_fix stuff.

reduzent commented:
> > I tried to avoid locks by switch off callbacks with each driver change with a message like params @callback 0, driver $1 but it still locks Pd.
> 
> This does not work because the `[driver ...(` message is still sent from within the audio callback. You need to add some delay so that Pd has a chance to exit the callback scheduler and enter the polling scheduler. Actually, a `[del 0]` between `[params @callback 0(` and `[driver ...(` already seems to do the trick. (Tested on Windows by switching from Portaudio + callbacks to MMIO.) Needless to say, this is just a workaround and not a real solution.

Ok, I didn't test with `[del 0]` between the `callback 0` and the `driver $1` message. That would be a totally viable solution for me. Before `scheduler_fix` the time required was totally non-deterministic and in the range of seconds, so it was really hard to not run into a lock from time to time. 

Now with 6b0d66b, Pd locks always when setting driver to JACK with `driver 5` to `[audiosettings]`. Switching backends through the menu is still fine, though. Tested on macOS and Linux.

Spacechild1 commented:
The problem is that `[audiosettings]` directly calls `sys_reopen_audio` from a Pd message, so Pd would be locked. `jack_open_audio` sets a callback when the sample rate changes. Strangely, that callback is invoked immediately. The samplerate callback locks Pd before updating the DSP graph - which will in turn dead lock because Pd is already locked.

Actually, now that I think of it, `sys_reopen_audio` and `sys_close_audio` should probably just set a flag and notify the scheduler, i.e. take the role of `sched_reopenaudio` and `sched_closeaudio`. The actual implementation could be moved to (private) helper functions. This would allow externals to safely call `sys_reopen_audio` and `sys_close_audio` inside Pd messages and the `[audiosettings]` object would work as expected. I will give it a shot.

Spacechild1 commented:
@reduzent With [e48d2e3](https://github.com/pure-data/pure-data/pull/1756/commits/e48d2e37d1df3520ab83e875ee9ff5606821b55f) the `[audiosettings]` object should work again. Now you can even change backends, with callbacks enabled, without any workarounds!

`sys_close_audio` and `sys_reopen_audio` are now 100% safe, as they just notify the scheduler. (They essentially replace `sched_reopenaudio` and `sched_closeaudio`.) The actual implementation has been moved to *private* functions `sys_do_reopen_audio` and `sys_do_close_audio`.

Maybe that's the better design after all: make the public functions safe and move the risky stuff into private functions.

reduzent commented:
> @reduzent With e48d2e3 the `[audiosettings]` object should work again. Now you can even change backends, with callbacks enabled, without any workarounds!

Sweet potato! Tested on Linux and it works smoothly through menu and `[audiosettings]`. That's great!

Spacechild1 commented:
> Sweet potato! Tested on Linux and it works smoothly through menu and [audiosettings]. That's great!

Yay! I have now integrated the last commit into the previous commits for a cleaner history. Looks like I'm done for now. Thanks for all the testing!

reduzent commented:
> > @reduzent With [e48d2e3](https://github.com/pure-data/pure-data/commit/e48d2e37d1df3520ab83e875ee9ff5606821b55f) the `[audiosettings]` object should work again. Now you can even change backends, with callbacks enabled, without any workarounds!

Now, I tested also on macOS and it works all well. While I write this, I see new commits being pushed, so maybe I should mention that I tested specifically e48d2e37.

@millerpuckette The handling of back-ends of this PR is a clear improvement over what we had until now. This is well tested. I hope this PR is considered for merging.

Spacechild1 commented:
> While I write this, I see new commits being pushed, so maybe I should mention that I tested specifically https://github.com/pure-data/pure-data/commit/e48d2e37d1df3520ab83e875ee9ff5606821b55f.

I just squashed some commits. I double checked and everything still works on my side, so it should be ok.

I think this PR is now ready for merging.

danomatika commented:
> Maybe that's the better design after all: make the public functions safe and move the risky stuff into private functions.

In general, I would say YES. ;)

Spacechild1 commented:
Aaand some more commits.

reduzent commented:
Now, this branch has conflicts with master. Hopefully, this gets merged soon and doesn't require multiple rounds of conflict fixes.

BTW: I've been using a Pd from this branch since March on Linux and macOS and I haven't encountered any new issues. It's a clear improvement, I can't stretch that enough.

millerpuckette commented:
I think these changes will be very useful... however, the original m_sched was largely system-independent, with things such as locking and thread management quarantined in (the very ugly) s_inter.c . This is important for people wanting to use libpd in bare-metal applications. An example is espd (msp.ucsd.edu/ideas/espd), which uses m_sched.c but not s_inter.c.

Perhaps it would be possible to move all the dreadful semaphores and system clock functions to a new file, perhaps s_audio_(something).c?

danomatika commented:
Maybe something like s_inter_sched.c with a common internal header? Then you could build Pd with an alternate dummy file with stubs for a minimal scheduler?

enohp ym morf tnes
-----------
Dan Wilcox
danomatika.com
robotcowboy.com


> On Jun 5, 2023, at 12:25 PM, Miller Puckette ***@***.***> wrote:
> 
> ﻿
> I think these changes will be very useful... however, the original m_sched was largely system-independent, with things such as locking and thread management quarantined in (the very ugly) s_inter.c . This is important for people wanting to use libpd in bare-metal applications. An example is espd (msp.ucsd.edu/ideas/espd), which uses m_sched.c but not s_inter.c.
> 
> Perhaps it would be possible to move all the dreadful semaphores and system clock functions to a new file, perhaps s_audio_(something).c?
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.

millerpuckette commented:
This is important but will probably take a while to work through so I suggest we plan on fixing and adopting this for 0.55 rather than 0.54.

Spacechild1 commented:
I have rebased on top of 0.54 and added a 0.55 milestone.

> Perhaps it would be possible to move all the dreadful semaphores and system clock functions to a new file, perhaps 

Regarding `m_sched.c` I think we only need to move the clock stuff from `m_sched.c` to a dedicated file (e.g. `m_clock.c`). Then we would only build `m_sched.c` for the actual Pd app. But can we do that as a second step *after* this PR has been merged? The change set is already pretty large...

[I think we should also think about breaking up `s_inter.c` as it contains a whole bag of unrelated functionality. For example, on bare metal I probably won't need the GUI communication code, but I still might want to register poll functions and do network I/O. There probably should be a larger discussion about restructuring some of the codebase for better modularity, but please not in this PR :)]

---

BTW, I just tested again today and here on Windows 10 I can now run Pd at 64 samples with 3 ms delay :)

reduzent commented:
I often turn DSP off during operations that would otherwise trigger multiple DSP graph calculations. It appears that with this branch, turning DSP off and on quickly enough makes Pd lose sound. DSP is still calculated, but soundcard is silent. Running another cycle of DSP off and on brings audio back.

This does not happen when using JACK, but it happens with the other backends. I tested:
  * Linux / ALSA
  * macOS / "standard (portaudio)" (CoreAudio)
  * Windows (I only got reports, didn't test myself)

Check attached patch. Make sure that audio is running and you hear the sine. When you hit the bang, audio is turned off. The [env~] indicates that DSP is still running. Another click on the bang, brings audio back. 

Expected behaviour: Audio is always on. 

[test-dsp-off-on.zip](https://github.com/pure-data/pure-data/files/13513608/test-dsp-off-on.zip)

Spacechild1 commented:
Good catch! Should be fixed with the latest commit. Please try with Jack and portaudio.

millerpuckette commented:
I'd like to adopt this but I simply disagree with not having quit mean really quit - I've seen too many situations where Pd hangs trying to quit cleanly.  (For example, patches with lots of copies of the same abstraction can take a long time to close; a workaround is simply to quit Pd without closing the patch.  I've also seen Pd hang trying to close audio and/or MIDI on various platforms.

Meanwhile, I want to change the startup order in Pd to open patches and handle "-send" startup args before waiting for the startup message from the GUI (glob_initfromgui) to fix an ugly race condition in pd~  - but I think that change will conflict with this PR.  What do you think of my trying to cherry-pick from this one (everything except the clean-shutdown bit) before trying to do that?

Spacechild1 commented:
> What do you think of my trying to cherry-pick from this one (everything except the clean-shutdown bit) before trying to do that?

I think manually cherry-picking all these commits is a bit risky :) If really needed, I can just remove that one commit and force push. But first consider the following:

>  I've also seen Pd hang trying to close audio and/or MIDI on various platforms.

Here's the original code for `glob_exit`:
```
void glob_exit(void *dummy, t_float status)
{
    sys_exit();
    sys_close_audio();
    sys_close_midi();
    if (sys_havegui())
    {
        sys_closesocket(INTER->i_guisock);
        sys_rmpollfn(INTER->i_guisock);
    }
    exit((int)status);
}
```
As you can see, this first closes the audio devices, then closes the GUI and finally exits.

This is logically equivalent to my commit. In fact, I think my code is *less* likely to hang with the callback scheduler, because it doesn't try to quit it on the audio thread (which may indeed block).

The only difference at the moment is that I also close all the parent canvases for a graceful shutdown. This is mainly done for 2 reasons:

1. by closing all canvases before exiting, we make sure that the external destructors are run and custom "closebang" actions are triggered. (For example, I have patches where the config is autosaved on closing.)

2. `pd_term_systems` is still empty, but the idea is that this is place where we properly free all global resources.
    * IMO Pd should not leak memory on shutdown, so we have a chance to catch *actual* memory leaks!
    * It would also allow us to properly free all the classes, which would also call the user-provided `t_classfreefn`; at the moment, `class_free` is never called!
       (The class free function would be very handy for external libraries that really need to be properly terminated; I have seen VST plugin DLLs segfault on exit because we have no chance to properly deinitialize it.)

>  For example, patches with lots of copies of the same abstraction can take a long time to close; a workaround is simply to quit Pd without closing the patch

That seems to be your actual issue. Personally, I don't like that closing the Pd app has a different behavior than subsequently closing all windows. This is just not very intuitive. See point 1. above.

I mean, if a patch takes a long time to close, than it probably also takes a lot of time to open anyway... This is seems to be a bit of an edge case, and I am not sure it outweights the gains. In theory, we could just add some force quit message and/or menu entry that bypasses the graceful shutdown.

If you *really* don't want Pd to close the canvases, I would suggest that I just wrap the relevant code in a `#if 0` clause, so at least I (and others) can easily activate it :)

millerpuckette commented:
The most recent problem I've had is that abstractions stupidly bind themselves to "pd-myabstraction" style symbols, and unbinding them is order n^2.  If you have tens of thousands of them this makes patches VERY slow to close.  I want to figure out a good fix for that someday.  But meanwhile, what if quitting Pd called all the closebangs but didn't actually take the time to close all the patches?  Or even better, what if that were the default and there were a startup flag or something to ask that patches really get closed (as you prefer)?

millerpuckette commented:
Meanwhile I'll email my current, not fully tested, s_main.c and s_inter.c so you can see if there are going to be conflicts... github doesn't want me to attach them here :)

Spacechild1 commented:
> The most recent problem I've had is that abstractions stupidly bind themselves to "pd-myabstraction" style symbols, and unbinding them is order n^2. If you have tens of thousands of them this makes patches VERY slow to close. 

:-D

> But meanwhile, what if quitting Pd called all the closebangs but didn't actually take the time to close all the patches? Or even better, what if that were the default and there were a startup flag or something to ask that patches really get closed (as you prefer)?

That's a good idea! I would suggest that I just change
```
void pd_term(void)
{
    t_glist *c;
    while ((c = pd_getcanvaslist()))
    {
        canvas_closebang(c);
        pd_free((t_pd *)c);
    }
    pd_term_systems();
}
```
to
```
void pd_term(void)
{
    t_glist *c;
    while ((c = pd_getcanvaslist()))
    {
        canvas_closebang(c);
#if 0
        pd_free((t_pd *)c);
#endif
    }
    pd_term_systems();
}
```
That would at least call the closebangs. LATER we can figure out how to handle the actual canvases. What do you think?

Spacechild1 commented:
>  I want to figure out a good fix for that someday.

I would be curious to know how slow it would be with https://github.com/pure-data/pure-data/pull/1760. This implementation uses arrays instead of linked lists for the bindlists, so I would expect unbinding to be much faster.

Spacechild1 commented:
I just updated `pd_term` to not free the canvases by default and force pushed.

Spacechild1 commented:
@millerpuckette Looks like it still merges cleanly, even after your recent changes. Let's take the opportunity :)

millerpuckette commented:
hmm... this breaks libpd - which depends on sched_tick always advancing by one DSP tick.  Is there a way to get this to work and preserve the fact that sched_tick always advances time?

Spacechild1 commented:
>  Is there a way to get this to work and preserve the fact that sched_tick always advances time?

I didn't touch `sched_tick` at all... What exactly isn't working?

millerpuckette commented:
I think it's that sys_quit keeps getting set - which causes sched_tick to bail without ticking.

Spacechild1 commented:
Good catch! I think I found it. Can you try with the latest commit?

Spacechild1 commented:
PS: the CI only fails because of a trailing whitespace in `s_inter.c` on current master :)

danomatika commented:
Thanks for checking libpd BTW. This reminds me to make a PR with the current libpd C sources for 0.55.enohp ym morf tnes-----------Dan Wilcoxdanomatika.comrobotcowboy.comOn Apr 8, 2024, at 4:26 PM, Christof Ressi ***@***.***> wrote:﻿
PS: the CI only fails because of a trailing whitespace in s_inter.c on current master :)

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

Spacechild1 commented:
Hmmm... I'll have a look.

Spacechild1 commented:
Ok, the crash with `-nosound` should be fixed. I also caught a stupid regression in my recent `pd_term` fixup that would keep Pd in an infinite loop 🤦 

Is the libpd issue resolved?

porres commented:
@Spacechild1  is this that thing you mentioned in that SC talk?

Spacechild1 commented:
@millerpuckette Do you see any more problems?

Spacechild1 commented:
@millerpuckette Awesome, thanks a lot!

danomatika commented:
Great! I think this helps with a lot of things.

@Spacechild1 Is there anything to change in lined for this?

Spacechild1 commented:
> @Spacechild1 Is there anything to change in libpd for this?

I don't think so!
